tty.c

```diff
@@ -109,101 +116,232 @@ void tty_setup_socketpair(int *pParentSock, int *pChildSock) {
 }
 
-bool tty_parent_select(pid_t pid) {
-	const size_t buflen = 1024;
-	char buf[buflen];
-	fd_set readFds = info.rfds, writeFds = info.wfds;
-	bool rtn = false;
+static struct epoll_handler inbound_handler, outbound_handler, term_handler;
 
-	int rc = select(info.nfds, &readFds, NULL, NULL, NULL);
-	if (rc == 0) {
-		return false;
-	}
-	if (rc < 0) {
-		if (errno == EINTR) {
-			return false;
+static struct buffer inbound_buffer, outbound_buffer;
+
+static int tty_handle_io(int epollfd, const struct epoll_event *ev, int fd, pid_t pid)
+{
+	struct epoll_handler *handler = ev->data.ptr;
+
+	if (fd == inbound_handler.fd) {
+		if (ev->events & EPOLLIN) {
+			handler->ready |= READ_READY;
 		}
-		err(1, "select");
-	}
-	struct timeval immediate = {0};
-	if (select(info.nfds, NULL, &writeFds, NULL, &immediate) < 0) {
-		return false;
-	}
-	if (FD_ISSET(STDIN_FILENO, &readFds) && FD_ISSET((unsigned long) info.termfd, &writeFds)) {
-		ssize_t nread = read(STDIN_FILENO, buf, buflen);
-		if (nread > 0) {
-			if (write(info.termfd, buf, (size_t) nread) < 0) {
-				warn("writing to terminal");
-			}
-		} else {
-			if (nread < 0) {
-				warn("reading from stdin");
-			}
-			FD_CLR(STDIN_FILENO, &info.rfds);
-			if (write(info.termfd, &(char){4}, 1) < 0) {
-				warn("writing EOT to terminal");
+		if (ev->events & EPOLLHUP) {
+			send_veof(handler->peer_fd);
+		}
+	} else if (fd == outbound_handler.fd) {
+		if (ev->events & EPOLLOUT) {
+			handler->ready |= WRITE_READY;
+		}
+	} else {
+		struct epoll_event newev = *ev;
+		newev.events = EPOLLIN | EPOLLOUT | EPOLLONESHOT;
+
+		if (ev->events & EPOLLOUT || inbound_handler.ready & WRITE_READY) {
+			inbound_handler.ready |= WRITE_READY;
+			newev.events &= ~EPOLLOUT;
+		}
+		if (ev->events & EPOLLIN || outbound_handler.ready & READ_READY) {
+			outbound_handler.ready |= READ_READY;
+			newev.events &= ~EPOLLIN;
+		}
+		if (!(ev->events & EPOLLHUP) && newev.events != EPOLLONESHOT) {
+			if (epoll_ctl(epollfd, EPOLL_CTL_MOD, fd, &newev) == -1) {
+				err(1, "epoll_ctl_mod termfd");
 			}
 		}
 	}
-	if (FD_ISSET((unsigned long) info.termfd, &readFds) && FD_ISSET(STDOUT_FILENO, &writeFds)) {
-		ssize_t nread = read(info.termfd, buf, buflen);
-		if (nread > 0) {
-			if (write(STDOUT_FILENO, buf, (size_t) nread) < 0) {
-				warn("writing to stdout");
-			}
-		} else {
-			if (nread < 0 && errno != EIO) {
-				warn("reading from terminal");
+
+	struct epoll_event termev = {
+		.events = EPOLLONESHOT,
+		.data.ptr = &term_handler,
+	};
+	int termfd = handler->fd;
+
+	if (inbound_handler.ready == (READ_READY | WRITE_READY)) {
+		int read_fd = inbound_handler.fd;
+		int write_fd = inbound_handler.peer_fd;
+		termfd = inbound_handler.peer_fd;
+		termev.events |= EPOLLOUT;
+
+		size_t remain = inbound_buffer.size;
+		ssize_t copied = io_copy(write_fd, read_fd, &inbound_buffer);
+		if (copied == -1) {
+			err(1, "copy stdin -> tty");
+		}
+		if (copied - remain == 0) {
+			if (write(termfd, &(char){4}, 1) < 0) {
```

Comment:
Is this the same as send_veof or different?

Reply:
same comment as above